x-pack/plugins/stack_connectors/server/connector_types/crowdstrike/crowdstrike.ts

```diff
@@ -0,0 +1,181 @@
+import type { AxiosError } from 'axios';
+import { SubActionRequestParams } from '@kbn/actions-plugin/server/sub_action_framework/types';
+import type {
+  CrowdstrikeConfig,
+  CrowdstrikeSecrets,
+  CrowdstrikeGetAgentsResponse,
+  CrowdstrikeGetAgentsParams,
+  CrowdstrikeBaseApiResponse,
+  CrowdstrikeHostActionsParams,
+  CrowdstrikeGetTokenResponse,
+} from '../../../common/crowdstrike/types';
+import {
+  CrowdstrikeGetAgentsResponseSchema,
+  CrowdstrikeHostActionsParamsSchema,
+  CrowdstrikeGetAgentsParamsSchema,
+  CrowdstrikeGetTokenResponseSchema,
+  CrowdstrikeHostActionsResponseSchema,
+} from '../../../common/crowdstrike/schema';
+import { SUB_ACTION } from '../../../common/crowdstrike/constants';
+
+export const API_PATH = 'https://api.crowdstrike.com';
+
+const paramsSerializer = (params: Record<string, string>) => {
+  return Object.entries(params)
+    .map(([key, value]) => `${encodeURIComponent(key)}=${encodeURIComponent(value)}`)
+    .join('&');
+};
+
+export class CrowdstrikeConnector extends SubActionConnector<
+  CrowdstrikeConfig,
+  CrowdstrikeSecrets
+> {
+  private static token: string | null;
+  private static tokenExpiryTimeout: NodeJS.Timeout;
+
+  private urls: {
+    getToken: string;
+    agents: string;
+    hostAction: string;
+  };
+
+  constructor(params: ServiceParams<CrowdstrikeConfig, CrowdstrikeSecrets>) {
+    super(params);
+    this.urls = {
+      getToken: `${API_PATH}/oauth2/token`,
+      hostAction: `${API_PATH}/devices/entities/devices-actions/v2`,
+      agents: `${API_PATH}/devices/entities/devices/v2`,
+    };
+
+    this.registerSubActions();
+  }
+
+  private registerSubActions() {
+    this.registerSubAction({
+      name: SUB_ACTION.GET_AGENT_DETAILS,
+      method: 'getAgentDetails',
+      schema: CrowdstrikeGetAgentsParamsSchema,
+    });
+
+    this.registerSubAction({
+      name: SUB_ACTION.HOST_ACTIONS,
+      method: 'executeHostActions',
+      schema: CrowdstrikeHostActionsParamsSchema,
+    });
+  }
+
+  public async executeHostActions({ alertIds, ...payload }: CrowdstrikeHostActionsParams) {
+    return this.crowdstrikeApiRequest({
+      url: this.urls.hostAction,
+      method: 'post',
+      params: {
+        action_name: payload.command,
+      },
+      data: {
+        ids: payload.ids,
+      },
+      paramsSerializer,
+      responseSchema: CrowdstrikeHostActionsResponseSchema,
+    });
+  }
+
+  public async getAgentDetails(
+    payload: CrowdstrikeGetAgentsParams
+  ): Promise<CrowdstrikeGetAgentsResponse> {
+    return this.crowdstrikeApiRequest({
+      url: this.urls.agents,
+      method: 'GET',
+      params: {
+        ids: payload.ids,
+      },
+      paramsSerializer,
+      responseSchema: CrowdstrikeGetAgentsResponseSchema,
+    });
+  }
+
+  private async getTokenRequest() {
+    const base64encodedData = Buffer.from(
+      this.secrets.clientId + ':' + this.secrets.clientSecret
+    ).toString('base64');
```

Comment:
I don't remember if this operation is costly or not, but regardless - it can probably just be done one and then reused on subquent calls to `getTokeRequest()`